axum-core/src/body.rs

```diff
@@ -182,8 +121,16 @@ impl Stream for Body {
     type Item = Result<Bytes, Error>;
 
     #[inline]
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        self.poll_data(cx)
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match futures_util::ready!(self.as_mut().poll_frame(cx)?) {
+                Some(frame) => match frame.into_data() {
+                    Ok(data) => return Poll::Ready(Some(Ok(data))),
+                    Err(_frame) => {}
```

Comment:
This is for streaming headers and trailers, right? Or just trailers? Is it really okay to discard these in the `Stream` impl? What is this impl even used for? :sweat_smile:

Reply:
Yeah `Err(_frame)` is if the body returns a non-data frame, which can currently only be trailers.

http-body-util provides [a stream adapter which streams frames](https://docs.rs/http-body-util/latest/http_body_util/struct.StreamBody.html). This impl just streams the bytes in data frames which I think is more generally useful. Makes things like streaming request bodies to files easier.

It is a bit iffy to ignore trailers but [`BodyStream` in 0.6](https://docs.rs/axum/latest/src/axum/extract/request_parts.rs.html#143-149) does it too so I think its okay.

Reply:
Been thinking a little more about this and I think adding a dedicated method to turn the body into a data stream is worth it. So I've added `Body::into_data_stream` and `Body` no longer implements `Stream` directly.